frontend/test/pytest/test_measurement_primitives.py

```diff
@@ -104,3 +159,20 @@ def f():
     assert jaxpr.eqns[1].primitive == state_p
     assert jaxpr.eqns[1].params == {"shape": (1,), "shots": 5}
     assert jaxpr.eqns[1].outvars[0].aval.shape == (1,)
+
+
+def test_state_dynamic():
+    """Test that the state primitive can be captured by jaxpr
+    when using tracers."""
+
+    def f(dim_0):
+        obs = compbasis_p.bind()
+        return state_p.bind(obs, shots=5, shape=(dim_0,))
+
+    jaxpr = jax.make_jaxpr(f)(1)
+    shots, shape_value = jaxpr.eqns[1].params.values()
```

Comment:
I think we still need to cover more of the pipeline in our tests (ignoring the MLIR core for now, which can be addressed in a different PR). 
From what I can see we test the bind function and abstract evaluation of jaxpr primitives, but that leaves out most of the code that constitutes the frontend. The first step might be to identify which functions could be affected by this change, from the start of tracing until jax spits out the mlir, and then ensure those functions are tested for the dynamic case. Or alternatively develop tests that cover roughly that entire span.